File: SimpleAdapter/src/main/java/by/shostko/android/adapter/simple/filter/SimpleFilter.java
package by.shostko.android.adapter.simple.filter;

import android.support.annotation.NonNull;
import android.support.v7.util.DiffUtil;

import java.util.List;

public class SimpleFilter<T, F> extends BaseFilter<T, Object, F>
{
    @NonNull
    private final FilterableHelper<T, F> filterableHelper;

    @NonNull
    private final DiffHelper<T> diffHelper;

    public SimpleFilter(@NonNull FilterableHelper<T, F> filterableHelper,
                        @NonNull DiffHelper<T> diffHelper,
                        @NonNull AdapterHelper<T, Object> adapterHelper)
    {
        super(filterableHelper, diffHelper, adapterHelper);
        this.filterableHelper = filterableHelper;
        this.diffHelper = diffHelper;
    }

    @Override
    protected boolean checkForAll(@NonNull F constraint, @NonNull Object param)
    {
        return filterableHelper.checkForAll(constraint);
    }

    @Override
    protected boolean check(@NonNull F constraint, @NonNull T item, @NonNull Object param)
    {
        return filterableHelper.check(constraint, item);
    }

    @NonNull
    @Override
    protected DiffUtil.Callback getDiffCallback(@NonNull List<T> oldList, @NonNull Object oldParam,
                                                @NonNull List<T> newList, @NonNull Object newParam)
    {
        return diffHelper.getDiffCallback(oldList, newList);
    }
}

File: SimpleAdapter/src/main/java/by/shostko/android/adapter/simple/v2/SimpleFilter.kt
package by.shostko.android.adapter.simple.v2

import android.support.v7.util.DiffUtil
import android.widget.Filter

class SimpleFilter<T, P, F> internal constructor(
        private val adapter: CoreAdapter<*, T, P, F>,
        private val original: Data<T, P>,
        private val previous: Data<T, P>,
        private val filterDelegate: FilterDelegate.Parametrized<T, P, F>,
        private var diffDelegate: DiffDelegate.Parametrized<T, P>
) : Filter() {

    override fun performFiltering(constraint: CharSequence?): FilterResults {
        val filter = filterDelegate.convert(constraint)
        val filtered = if (filterDelegate.checkAll(original.param, filter)) {
            original.items
        } else {
            original.items.filterIndexed { position, item ->
                val viewType = diffDelegate.getItemViewType(position, item, original.param)
                filterDelegate.check(position, viewType, item, original.param, filter)
            }
        }

        val diffCallback = SimpleDiff(previous, Data(filtered, original.param), diffDelegate)
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        val results = Filter.FilterResults()
        results.values = Pair(filtered, diffResult)
        results.count = filtered.size
        return results
    }

    @Suppress("UNCHECKED_CAST")
    override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
        val (filtered, diffResult) = results?.values as Pair<Data<T, P>, DiffUtil.DiffResult>
        adapter.publishResults(original, filtered, diffResult)
    }
}